polkadot/xcm/xcm-builder/src/fee_handling.rs

```diff
@@ -17,42 +17,87 @@
 use core::marker::PhantomData;
 use frame_support::traits::{Contains, Get};
 use xcm::prelude::*;
-use xcm_executor::traits::{FeeManager, FeeReason, TransactAsset};
+use xcm_executor::traits::{FeeManager, FeeReason, HandleFee, TransactAsset};
 
-/// A `FeeManager` implementation that simply deposits the fees handled into a specific on-chain
+/// A `FeeManager` implementation that permits the specified `WaivedLocations` to not pay for fees
+/// and that uses the provided `HandleFee` implementation otherwise.
+pub struct XcmFeeManagerFromComponents<WaivedLocations, HandleFee>(
+	PhantomData<(WaivedLocations, HandleFee)>,
+);
+impl<WaivedLocations: Contains<MultiLocation>, FeeHandler: HandleFee> FeeManager
+	for XcmFeeManagerFromComponents<WaivedLocations, FeeHandler>
+{
+	type HandleFee = FeeHandler;
+
+	fn is_waived(origin: Option<&MultiLocation>, _: FeeReason) -> bool {
+		let Some(loc) = origin else { return false };
+		WaivedLocations::contains(loc)
+	}
+}
+
+/// A `HandleFee` implementation that simply deposits the fees into a specific on-chain
 /// `ReceiverAccount`.
 ///
-/// It reuses the `AssetTransactor` configured on the XCM executor to deposit fee assets, and also
-/// permits specifying `WaivedLocations` for locations that are privileged to not pay for fees. If
+/// It reuses the `AssetTransactor` configured on the XCM executor to deposit fee assets. If
 /// the `AssetTransactor` returns an error while calling `deposit_asset`, then a warning will be
 /// logged.
-pub struct XcmFeesToAccount<XcmConfig, WaivedLocations, AccountId, ReceiverAccount>(
-	PhantomData<(XcmConfig, WaivedLocations, AccountId, ReceiverAccount)>,
+pub struct XcmFeeToAccount<AssetTransactor, AccountId, ReceiverAccount>(
+	PhantomData<(AssetTransactor, AccountId, ReceiverAccount)>,
 );
+
 impl<
-		XcmConfig: xcm_executor::Config,
-		WaivedLocations: Contains<MultiLocation>,
+		AssetTransactor: TransactAsset,
 		AccountId: Clone + Into<[u8; 32]>,
-		ReceiverAccount: Get<Option<AccountId>>,
-	> FeeManager for XcmFeesToAccount<XcmConfig, WaivedLocations, AccountId, ReceiverAccount>
+		ReceiverAccount: Get<AccountId>,
```

Comment:
Do not know if there was some design solution behind having `Get<Option<AccountId>>` in the previous version? Maybe some storage parameter was intended to be used - so fees still may be burnt unless governance has configured chain properly? But in any case - new version looks cleaner to me, just stumbled upon it

Reply:
I think `Get<Option<AccountId>>` was needed in order to support the case where we wanted to waive the fees for some locations and burn the ones for the other locations using the `XcmFeeToAccount`.

But now, that we have a separate component for handling fees, we can use `XcmFeeManagerFromComponents` to specify the set of waived locations and the handling logic independently. And we can use `XcmFeeManagerFromComponents<WaivedLocations, ()>` for the scenario above.